RESTEASY-1976 - Make tracing info to be returned in JSON format so it…

liweinan opened the pull request:
… can be better processed by consumers.

Additional details in https://github.com/resteasy/Resteasy/pull/1624

liweinan commented:
```
$ curl -i -H "X-RESTEasy-Tracing-Accept-Format: JSON" http://localhost:8081/type
HTTP/1.1 200 OK
Connection: keep-alive
X-RESTEasy-Tracing-000: [{"duration":0,"event":"START","timestamp":9421067942315},{"duration":0,"event":"START_HEADERS","timestamp":9421068723325},{"duration":20516,"event":"PRE_MATCH_SUMMARY","timestamp":9421070802178},{"duration":22644,"event":"REQUEST_FILTER","timestamp":9421073836502},{"duration":24271,"event":"REQUEST_FILTER","timestamp":9421074185191},{"duration":761813,"event":"REQUEST_FILTER_SUMMARY","timestamp":9421074379641},{"duration":0,"event":"MATCH_RUNTIME_RESOURCE","timestamp":9421074951355},{"duration":0,"event":"MATCH_PATH_FIND","timestamp":9421075122971},{"duration":0,"event":"MATCH_PATH_SELECTED","timestamp":9421076738564},{"duration":2247416,"event":"MATCH_SUMMARY","timestamp":9421076931578},{"duration":0,"event":"MATCH_RESOURCE","timestamp":9421077128495},{"duration":0,"event":"MATCH_RESOURCE_METHOD","timestamp":9421077327927},{"duration":469922,"event":"REQUEST_FILTER","timestamp":9421078972695},{"duration":813045,"event":"REQUEST_FILTER_SUMMARY","timestamp":9421079286322},{"duration":8355474,"event":"METHOD_INVOKE","timestamp":9421086259265},{"duration":0,"event":"DISPATCH_RESPONSE","timestamp":9421086621288},{"duration":11287,"event":"RESPONSE_FILTER","timestamp":9421093091764},{"duration":0,"event":"MBW_FIND","timestamp":9421095655441},{"duration":0,"event":"MBW_SELECTED","timestamp":9421096141575},{"duration":0,"event":"MBW_SKIPPED","timestamp":9421096346431},{"duration":0,"event":"MBW_SKIPPED","timestamp":9421096501493},{"duration":0,"event":"MBW_SKIPPED","timestamp":9421096654748},{"duration":0,"event":"MBW_SKIPPED","timestamp":9421096837082},{"duration":0,"event":"MBW_SKIPPED","timestamp":9421097010572},{"duration":0,"event":"MBW_SKIPPED","timestamp":9421097142411},{"duration":0,"event":"MBW_SKIPPED","timestamp":9421097303761},{"duration":0,"event":"MBW_SKIPPED","timestamp":9421097446038},{"duration":0,"event":"MBW_WRITE_TO","timestamp":9421097580849},{"duration":4167082,"event":"WI_AFTER","timestamp":9421099257737},{"duration":5088628,"event":"WI_SUMMARY","timestamp":9421099501441},{"duration":6623265,"event":"RESPONSE_FILTER_SUMMARY","timestamp":9421099683931},{"duration":0,"event":"FINISHED","timestamp":9421099829754}]
Content-Type: application/octet-stream
Content-Length: 3
Date: Tue, 31 Jul 2018 16:57:04 GMT

ALL$
```

liweinan commented:
This is related with https://github.com/resteasy/resteasy-extensions/pull/2/files

liweinan commented:
@asoldano Could you please teach me how to connect this PR with the changes in `resteasy-extensions`? Or the test won't pass.

asoldano commented on resteasy-jaxrs/src/main/java/org/jboss/resteasy/tracing/RESTEasyTracingUtils.java:
```diff
@@ -100,4 +100,9 @@ static RESTEasyTracingLevel getRESTEasyTracingThreshold(Configuration configurat
     }
 
 
+    public static String getTracingInfoFormat(HttpRequest request) {
```
This would be better as package protected instead of public.

asoldano commented:
@liweinan as for connecting the PRs... there's not much to do; if a PR relies on another one in resteasy-extensions, that has to be merged first (or better, a snapshot artifacts including the changes has to be uploaded to the repo)

liweinan commented:
@asoldano could you tell me how to do this:
a snapshot artifacts including the changes has to be uploaded to the repo

liweinan commented:
@asoldano Ah okay, I understand the whole workflow. Thanks sir!

liweinan commented:
add tests and find this:

```
Caused by: java.lang.ClassNotFoundException: org.eclipse.yasson.JsonBindingProvider from [Module "javax.json.bind.api" from local module loader @28f67ac7 (finder: local module finder @256216b3 (roots: /Users/weli/projs/resteasy/resteasy-upstream/testsuite/integration-tests/target/test-server/wildfly-12.0.0.Final/modules,/Users/weli/projs/resteasy/resteasy-upstream/testsuite/integration-tests/target/test-server/wildfly-12.0.0.Final/modules/system/layers/base))]
```

liweinan commented:
reference about json-b binding in wildfly:

https://developer.jboss.org/thread/277353

liweinan commented:
After setting up the `module.xml` correctly for `javax.json.bind.api`:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<module xmlns="urn:jboss:module:1.5" name="javax.json.bind.api">

    <properties>
        <property name="jboss.api" value="public"/>
    </properties>

    <resources>
        <resource-root path="javax.json.bind-api-1.0.jar"/>
        <!-- Insert resources here -->
    </resources>

    <dependencies>
        <module name="javax.json.api"/>
        <module name="org.eclipse.yasson" services="import"/>
    </dependencies>
</module>
```

The test get work:

```
[standalone@localhost:9990 /] deploy /tmp/war_on_demand_tracing.war
[standalone@localhost:9990 /]
```

```
$ curl -i -H "X-RESTEasy-Tracing-Accept: foo" -H "X-RESTEasy-Tracing-Accept-Format: JSON"  http://127.0.0.1:8080/war_on_demand_tracing/logger
HTTP/1.1 200 OK
Connection: keep-alive
X-Powered-By: Undertow/1
Server: WildFly/11
X-RESTEasy-Tracing-000: [{"duration":0,"event":"START","text":"baseUri=[http://127.0.0.1:8080/war_on_demand_tracing/] requestUri=[http://127.0.0.1:8080/war_on_demand_tracing/logger] method=[GET] authScheme=[n/a] accept=*/* accept-encoding=n/a accept-charset=n/a accept-language=n/a content-type=n/a content-length=n/a ","timestamp":180514322005745},{"duration":6982,"event":"PRE_MATCH_SUMMARY","text":"PreMatchRequest summary: 0 filters","timestamp":180514322386677},{"duration":5690,"event":"REQUEST_FILTER","text":"Filter by [org.jboss.resteasy.test.tracing.HttpMethodOverride @74c69469]","timestamp":180514323005314},{"duration":2986,"event":"REQUEST_FILTER","text":"Filter by [org.jboss.resteasy.core.AcceptHeaderByFileSuffixFilter @2e57c4]","timestamp":180514323026894},{"duration":65766,"event":"REQUEST_FILTER_SUMMARY","text":"Request summary: 2 filters","timestamp":180514323033068},{"duration":0,"event":"MATCH_RUNTIME_RESOURCE","text":"Matched resource: template=[[org.jboss.resteasy.core.registry.ClassExpression @14be594a]] regexp=[\\Q\\E(.*)] matches=[[org.jboss.resteasy.core.registry.SegmentNode @7cd6da03]] from=[]","timestamp":180514323149275},{"duration":0,"event":"MATCH_PATH_FIND","text":"Matching path [/logger]","timestamp":180514323161384},{"duration":0,"event":"MATCH_PATH_SELECTED","text":"Pattern [\\Qlogger\\E] IS selected","timestamp":180514323232899},{"duration":131600,"event":"MATCH_SUMMARY","text":"RequestMatching summary","timestamp":180514323239718},{"duration":0,"event":"MATCH_RESOURCE","text":"Resource instance: [org.jboss.resteasy.core.ResourceMethodInvoker @3cf36c2c]","timestamp":180514323312596},{"duration":0,"event":"MATCH_RESOURCE_METHOD","text":"Matched method  : public java.lang.String org.jboss.resteasy.test.tracing.TracingConfigResource.logger(org.jboss.resteasy.spi.HttpRequest) throws java.lang.NoSuchMethodException","timestamp":180514323360835},{"duration":2332,"event":"REQUEST_FILTER","text":"Filter by [org.jboss.resteasy.plugins.providers.jackson.PatchMethodFilter @4e118c01]","timestamp":180514323393832},{"duration":2909,"event":"REQUEST_FILTER","text":"Filter by [org.jboss.resteasy.plugins.providers.sse.SseEventSinkInterceptor @2803f880 #2147483647]","timestamp":180514323425419},{"duration":43151,"event":"REQUEST_FILTER_SUMMARY","text":"Request summary: 2 filters","timestamp":180514323431577},{"duration":266251,"event":"METHOD_INVOKE","text":"Resource [SINGLETON|class org.jboss.resteasy.test.tracing.TracingConfigResource|org.jboss.resteasy.test.tracing.TracingConfigResource@3fd5940] method=[public java.lang.String org.jboss.resteasy.test.tracing.TracingConfigResource.logger(org.jboss.resteasy.spi.HttpRequest) throws java.lang.NoSuchMethodException]","timestamp":180514323654636},{"duration":0,"event":"DISPATCH_RESPONSE","text":"Response: [org.jboss.resteasy.specimpl.BuiltResponse @62cf3941 <200/SUCCESSFUL|OK|java.lang.String @45fabb38>]","timestamp":180514324493569},{"duration":2723,"event":"RESPONSE_FILTER","text":"Filter by [org.jboss.resteasy.security.doseta.DigitalSigningInterceptor @7494602]","timestamp":180514324661939},{"duration":1546,"event":"RESPONSE_FILTER","text":"Filter by [org.jboss.resteasy.plugins.interceptors.MessageSanitizerContainerResponseFilter @30b96993 #4000]","timestamp":180514324677543},{"duration":16866,"event":"WI_BEFORE","text":"[org.jboss.resteasy.plugins.interceptors.GZIPEncodingInterceptor @76202ef5 #4000] BEFORE context.proceed()","timestamp":180514324853880},{"duration":0,"event":"MBW_FIND","text":"Find MBW for type=[java.lang.String] genericType=[java.lang.String] mediaType=[[javax.ws.rs.core.MediaType @dae2453]] annotations=[@javax.ws.rs.GET(), @javax.ws.rs.Path(value=/logger)]","timestamp":180514324952505},{"duration":0,"event":"MBW_SELECTED","text":"[org.jboss.resteasy.plugins.providers.StringTextStar @391fd80c] IS writeable","timestamp":180514324967925},{"duration":0,"event":"MBW_WRITE_TO","text":"WriteTo by org.jboss.resteasy.plugins.providers.StringTextStar","timestamp":180514324973425},{"duration":156462,"event":"WI_AFTER","text":"[org.jboss.resteasy.security.doseta.DigitalSigningInterceptor @1cc7b6f9] AFTER context.proceed()","timestamp":180514325021620},{"duration":1404,"event":"WI_AFTER","text":"[org.jboss.resteasy.plugins.interceptors.GZIPEncodingInterceptor @76202ef5 #4000] AFTER context.proceed()","timestamp":180514325030101},{"duration":333727,"event":"WI_SUMMARY","text":"WriteTo summary: 2 interceptors","timestamp":180514325035481},{"duration":469425,"event":"RESPONSE_FILTER_SUMMARY","text":"Response summary: 2 filters","timestamp":180514325125873},{"duration":0,"event":"FINISHED","text":"Response status: 200","timestamp":180514325200756}]
Content-Type: application/octet-stream
Content-Length: 48
Date: Wed, 01 Aug 2018 10:31:40 GMT

org.jboss.resteasy.tracing.RESTEasyTracingLogger$
```

It seems Wildfly 12 and 12- has this module setting problem. Let me find a way to overcome it for testing.
```

liweinan commented:
Okay the maven ant-run plugin can deal with `module.xml` and fix the broken dependency of json-b binding in some versions of Wildfly. The tests passed on my local machine now:

<img width="1753" alt="image" src="https://user-images.githubusercontent.com/201907/43517895-d48606ee-95bc-11e8-8848-88e692a8fd71.png">

liweinan commented on testsuite/integration-tests/pom.xml:
```diff
@@ -81,6 +83,12 @@
+                                <replace
+                                        file="${project.build.directory}/test-server/wildfly-${server.version}/modules/system/layers/base/javax/json/bind/api/main/module.xml">
+                                    <replacetoken><![CDATA[</dependencies>]]></replacetoken>
+                                    <replacevalue><![CDATA[<module name="org.eclipse.yasson" services="import"/>
```
Fix the jsonb binding module dependency, or it won't work correctly.

liweinan commented:
@asoldano Thank you for advice!

I prefer the first solution: Let `JsonFormattedRESTEasyTracingInfo` detect the jackson or jsonb automatically. Let me work on this.

asoldano commented:
@liweinan or maybe simply go with Jackson2, unless  there's a reason for preferring jsonb

liweinan commented:
@asoldano yes sir!

Anyway let me add a JSONB option anyway, but prefer Jackson2 as default.

liweinan commented:
Design:

<img width="935" alt="image" src="https://user-images.githubusercontent.com/201907/43534017-e45516d8-95e8-11e8-83c9-5504d4dae6c1.png">

liweinan commented:
But it can easily work under standalone environment:

```
$ curl -i -H "X-RESTEasy-Tracing-Accept: foo" -H "X-RESTEasy-Tracing-Accept-Format: JSON"  http://127.0.0.1:8081/level
HTTP/1.1 200 OK
Connection: keep-alive
X-RESTEasy-Tracing-000: [{"event":"START","duration":0,"timestamp":205555619637592,"text":"baseUri=[http://127.0.0.1:8081/] requestUri=[http://127.0.0.1:8081/level] method=[GET] authScheme=[n/a] accept=*/* accept-encoding=n/a accept-charset=n/a accept-language=n/a content-type=n/a content-length=n/a "},{"event":"START_HEADERS","duration":0,"timestamp":205555620207064,"text":"Other request headers: Accept=[*/*] Host=[127.0.0.1:8081] User-Agent=[curl/7.58.0] X-RESTEasy-Tracing-Accept=[foo] X-RESTEasy-Tracing-Accept-Format=[JSON] "},{"event":"PRE_MATCH_SUMMARY","duration":12727,"timestamp":205555622135034,"text":"PreMatchRequest summary: 0 filters"},{"event":"REQUEST_FILTER","duration":15137,"timestamp":205555623872820,"text":"Filter by [io.weli.tracing.HttpMethodOverride @3eb97a97]"},{"event":"REQUEST_FILTER","duration":17349,"timestamp":205555624105435,"text":"Filter by [org.jboss.resteasy.core.AcceptHeaderByFileSuffixFilter @1dbf739a]"},{"event":"REQUEST_FILTER_SUMMARY","duration":563493,"timestamp":205555624234132,"text":"Request summary: 2 filters"},{"event":"MATCH_RUNTIME_RESOURCE","duration":0,"timestamp":205555624659712,"text":"Matched resource: template=[[org.jboss.resteasy.core.registry.ClassExpression @2fe41179]] regexp=[\\Q\\E(.*)] matches=[[org.jboss.resteasy.core.registry.SegmentNode @1ad3a645]] from=[]"},{"event":"MATCH_PATH_FIND","duration":0,"timestamp":205555624802815,"text":"Matching path [/level]"},{"event":"MATCH_PATH_SELECTED","duration":0,"timestamp":205555626358384,"text":"Pattern [\\Qlevel\\E] IS selected"},{"event":"MATCH_SUMMARY","duration":2063871,"timestamp":205555626532319,"text":"RequestMatching summary"},{"event":"MATCH_RESOURCE","duration":0,"timestamp":205555626678731,"text":"Resource instance: [org.jboss.resteasy.core.ResourceMethodInvoker @790204b7]"},{"event":"MATCH_RESOURCE_METHOD","duration":0,"timestamp":205555626797272,"text":"Matched method  : public java.lang.String io.weli.tracing.TracingConfigResource.level(org.jboss.resteasy.spi.ResteasyDeployment)"},{"event":"REQUEST_FILTER","duration":364869,"timestamp":205555628419970,"text":"Filter by [org.jboss.resteasy.plugins.providers.sse.SseEventSinkInterceptor @708f2f1 #2147483647]"},{"event":"REQUEST_FILTER_SUMMARY","duration":566530,"timestamp":205555628599544,"text":"Request summary: 1 filters"},{"event":"METHOD_INVOKE","duration":6274977,"timestamp":205555633597777,"text":"Resource [SINGLETON|class io.weli.tracing.TracingConfigResource|io.weli.tracing.TracingConfigResource@6fe7aac8] method=[public java.lang.String io.weli.tracing.TracingConfigResource.level(org.jboss.resteasy.spi.ResteasyDeployment)]"},{"event":"DISPATCH_RESPONSE","duration":0,"timestamp":205555633935377,"text":"Response: [org.jboss.resteasy.specimpl.BuiltResponse @4dd2c596 <200/SUCCESSFUL|OK|java.lang.String @25bf46e8>]"},{"event":"RESPONSE_FILTER","duration":10617,"timestamp":205555639815037,"text":"Filter by [org.jboss.resteasy.plugins.interceptors.MessageSanitizerContainerResponseFilter @1ddfc4e1 #4000]"},{"event":"MBW_FIND","duration":0,"timestamp":205555641570782,"text":"Find MBW for type=[java.lang.String] genericType=[java.lang.String] mediaType=[[javax.ws.rs.core.MediaType @4ed94b44]] annotations=[@javax.ws.rs.GET(), @javax.ws.rs.Path(value=/level)]"},{"event":"MBW_SELECTED","duration":0,"timestamp":205555641756155,"text":"[org.jboss.resteasy.plugins.providers.StringTextStar @7d3052c9] IS writeable"},{"event":"MBW_SKIPPED","duration":0,"timestamp":205555641843364,"text":"[org.jboss.resteasy.plugins.providers.sse.SseEventOutputProvider @3dc3c833] is skipped"},{"event":"MBW_SKIPPED","duration":0,"timestamp":205555641920508,"text":"[org.jboss.resteasy.plugins.providers.FileProvider @5491fcd] is skipped"},{"event":"MBW_SKIPPED","duration":0,"timestamp":205555641996013,"text":"[org.jboss.resteasy.plugins.providers.ByteArrayProvider @761afa3] is skipped"},{"event":"MBW_SKIPPED","duration":0,"timestamp":205555642071770,"text":"[org.jboss.resteasy.plugins.providers.StreamingOutputProvider @3e64760] is skipped"},{"event":"MBW_SKIPPED","duration":0,"timestamp":205555642147721,"text":"[org.jboss.resteasy.plugins.providers.ReaderProvider @2047957f] is skipped"},{"event":"MBW_SKIPPED","duration":0,"timestamp":205555642228916,"text":"[org.jboss.resteasy.plugins.providers.DataSourceProvider @6dbd51cd] is skipped"},{"event":"MBW_SKIPPED","duration":0,"timestamp":205555642319116,"text":"[org.jboss.resteasy.plugins.providers.FileRangeWriter @67abf76e] is skipped"},{"event":"MBW_SKIPPED","duration":0,"timestamp":205555642415297,"text":"[org.jboss.resteasy.plugins.providers.InputStreamProvider @564761bc] is skipped"},{"event":"MBW_WRITE_TO","duration":0,"timestamp":205555642678432,"text":"WriteTo by org.jboss.resteasy.plugins.providers.StringTextStar"},{"event":"WI_AFTER","duration":3902377,"timestamp":205555645201008,"text":"[org.jboss.resteasy.plugins.interceptors.GZIPEncodingInterceptor @76908cc0 #4000] AFTER context.proceed()"},{"event":"WI_SUMMARY","duration":4527045,"timestamp":205555645433391,"text":"WriteTo summary: 1 interceptors"},{"event":"RESPONSE_FILTER_SUMMARY","duration":5778240,"timestamp":205555645560510,"text":"Response summary: 1 filters"},{"event":"FINISHED","duration":0,"timestamp":205555645684250,"text":"Response status: 200"}]
Content-Type: application/octet-stream
Content-Length: 7
Date: Wed, 01 Aug 2018 17:29:01 GMT

VERBOSE$
```

Because the classloading scheme is different from `jboss-module` environment.

liweinan commented:
I run the test with Wildfly 13 and it works:

<img width="866" alt="image" src="https://user-images.githubusercontent.com/201907/43543131-9c848dd4-9601-11e8-9824-c14c3c063f56.png">

lower than 13 it won't work.

So I guess this should be  a 13+ feature. Let me start to create `ExpectedFailingOnWildFly11/12`

liweinan commented:
In conclusion:

- In standalone environment, either Jackson2 or JSON-B can be adopted automatically.
- In wildfly environment, because we don't have `jackson` support in `resteasy-jaxrs` scope, so the json-b solution is selected.
- In addition, the Wildfly below version 13 doesn't have json-b supported correctly, so this feature won't work.

asoldano commented:
@liweinan , while I'm fine with this being a WFLY13+ feature (but in that case I would simply use jsonb only), I have two comments / thoughts:
1) how about we add dependency to yasson in resteasy-extension/resteasy-tracing-api and move the JsonFormattedRESTEasyTracingInfo.java there? resteasy-jaxrs hasn't had an explicit module dependency to jsonb/jackson so far, I'd rather avoid forcing it because of this
2) this look like a scenario in which the thread context classloader should be used for loading the mappers, instead of the defining classloader (which on wildfly is clearly the resteasy-jaxrs module and hence has no visibility on the jackson2 module). I suggest you try using that, it should be set to the deployment classloader, which has visibility on the resteasy modules that are on by default on wildfly (jackson2 included). On standalone, the threadcontext classloader and the defining classloader will be the same.

liweinan commented:
@asoldano Thanks for advices! Let me:

1. Move JsonFormattedRESTEasyTracingInfo into extension.
2. Try the different classloaders and make Jackson adoption work firstly. So we can get rid of any hacks on JSON-B or Yasson inside Wildfly. And so this feature could work on Wildfly 13-.

liweinan commented:
- Moved `JsonFormattedRESTEasyTracingInfo` to: https://github.com/resteasy/resteasy-extensions/pull/3 
- Removed `yasson` from `resteasy-jaxrs` JBoss Module.

Next step is to solve the classloading problem.

liweinan commented:
tests passed on Wildfly11, 12, 13 and standalone environment, so let me remove the unnecessary test groups.

liweinan commented:
@asoldano Before I do the cleanup of the PR, could you please help to review the current solution?

liweinan commented:
The new design will load the tracing info class by using ServiceLoader:

![607341534402671_ pic_hd](https://user-images.githubusercontent.com/201907/44193935-1c1a7700-a167-11e8-87f3-cb5c0a92a9e7.jpg)

On server side, users can provide the tracing info classes in `META-INF`:

![image](https://user-images.githubusercontent.com/201907/44193860-dfe71680-a166-11e8-8a16-3a99d365d05c.png)

```
org.jboss.resteasy.tracing.api.providers.TextBasedRESTEasyTracingInfo
org.jboss.resteasy.tracing.api.providers.Jackson2JsonFormatRESTEasyTracingInfo
org.jboss.resteasy.tracing.api.providers.JSONBJsonFormatRESTEasyTracingInfo
```

To use one of the above tracing info provider, user just need to add relative dependency in pom:

```
        <dependency>
            <groupId>org.jboss.resteasy</groupId>
            <artifactId>resteasy-tracing-api</artifactId>
        </dependency>
```

The above will provide the default text based tracing info. To enable Jackson2 or JSON-B support, add one of the following dependencies will be fine:

```
        <dependency>
            <groupId>org.jboss.resteasy</groupId>
            <artifactId>resteasy-jackson2-provider</artifactId>
            <version>${version.resteasy.testsuite}</version>
            <scope>provided</scope>
        </dependency>
```

```
        <dependency>
            <groupId>org.jboss.resteasy</groupId>
            <artifactId>resteasy-json-binding-provider</artifactId>
            <version>${version.resteasy.testsuite}</version>
            <scope>provided</scope>
        </dependency>
```

In Wildfly or EAP environment, these providers are loaded during Wildfly startup, so it will be fine.

liweinan commented:
@asoldano refactor done, and tests passed locally. Could you please help to review this?

asoldano commented on resteasy-jaxrs/src/main/java/org/jboss/resteasy/tracing/RESTEasyTracingLoggerImpl.java:
```diff
@@ -1,13 +1,7 @@
-import org.jboss.resteasy.tracing.api.RESTEasyTracingInfo;
-import org.jboss.resteasy.tracing.api.RESTEasyTracingLevel;
-import org.jboss.resteasy.tracing.api.RESTEasyTracingMessage;
+import org.jboss.resteasy.tracing.api.*;
```
Can you please avoid using * imports? It's more clear to know what exactly is used, especially given the stuff here comes from an optional dependency. Moreover, a checkstyle will soon be used and it will flag * imports.

asoldano commented on testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/tracing/BasicTracingTest.java:
```diff
@@ -48,7 +50,7 @@ public void testPresencesOfServerTracingEvents() {
     @OperateOnDeployment(WAR_BASIC_TRACING_FILE)
     public void testBasic() throws InterruptedException {
-//        war.as(ZipExporter.class).exportTo(new File("/tmp/" + war.getName()), true);
+        basic_war.as(ZipExporter.class).exportTo(new File("/tmp/" + basic_war.getName()), true);
```
What's this for / is it needed?

liweinan commented:
Thanks for checking sir! Let me update this PR.

liweinan commented:
Now the services file inside `META-INF` is included into relative providers. So users don't have to create the service file by themselves.

liweinan commented:
All tests passed locally. @asoldano Could you please help to review this PR and relative one on `resteasy-extension` side? Thanks!
